RELEASE/scripts/autoscend/auto_quest_level_13.ash

```diff
@@ -1,5 +1,214 @@
+	{
+		return false;
+	}
+	if(item_amount($item[Digital Key]) > 0)
+	{
+		return false;
+	}
+	if(whitePixelCount() >= 30)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+int whitePixelCount()
+{
+	int count = item_amount($item[White Pixel]);
+
+	int extra = min(item_amount($item[Red Pixel]), item_amount($item[Blue Pixel]));
+	extra = min(extra, item_amount($item[Green Pixel]));
+	return count + extra;
+}
+
+boolean LX_getStarKey()
+{
+	if(!get_property("auto_getStarKey").to_boolean())
+	{
+		return false;
+	}
+
+	//needStarKey() checks if you own or have used the star key
+	if(!needStarKey())
+	{
+		set_property("auto_getStarKey", false);
+		return false;
+	}
+	
+	boolean hole_in_sky_unreachable = internalQuestStatus("questL10Garbage") < 9;
+	boolean shen_might_request_hole = internalQuestStatus("questL11Shen") < 7;
+	if (hole_in_sky_unreachable || shen_might_request_hole)
+	{
+		return false;
+	}
+	
+	//kingdom of exploathing does not need rocketship to reach hole in the sky
+	if(item_amount($item[Steam-Powered Model Rocketship]) == 0 && !in_koe())
+	{
+		return false;
+	}
+	
+	boolean at_tower_door = internalQuestStatus("questL13Final") == 5;
+	if (!in_hardcore() && at_tower_door && item_amount($item[Richard\'s Star Key]) == 0 && item_amount($item[Star Chart]) == 0 && !get_property("nsTowerDoorKeysUsed").contains_text("Richard's star key"))
+	{
+		pullXWhenHaveY($item[Star Chart], 1, 0);
+	}
+	
+	if((item_amount($item[Richard\'s Star Key]) == 0) && (item_amount($item[Star Chart]) > 0) && (item_amount($item[star]) >= 8) && (item_amount($item[line]) >= 7) && !get_property("nsTowerDoorKeysUsed").contains_text("Richard's star key"))
+	{
+		return create(1, $item[Richard\'s Star Key]);
+	}
+	
+	//if only star chart is missing and you will pull it at tower door, then you are done for now.
+	if((item_amount($item[Star]) >= 8) && (item_amount($item[Line]) >= 7) && !in_hardcore() && !at_tower_door)
+	{
+		return false;
+	}
+	
+	if(!zone_isAvailable($location[The Hole In The Sky]))
+	{
+		auto_log_warning("The Hole In The Sky is not available, we have to do something else...", "red");
+		return false;
+	}
+
+	if(auto_have_familiar($familiar[Space Jellyfish]))
+	{
+		handleFamiliar($familiar[Space Jellyfish]);
+		if(item_amount($item[Star Chart]) == 0)
+		{
+			set_property("choiceAdventure1221", 1);
+		}
+		else
+		{
+			set_property("choiceAdventure1221", 2 + (my_ascensions() % 2));
+		}
+	}
+	else
+	{
+		handleFamiliar("item");
+	}
+	
+	return autoAdv(1, $location[The Hole In The Sky]);
+}
+
+boolean LX_getDigitalKey()
+{
+	//Acquire the [Digital Key]
+	
+	if(contains_text(get_property("nsTowerDoorKeysUsed"), "digital key"))
+	{
+		return false;
+	}
+	if(item_amount($item[Digital Key]) > 0)
+	{
+		if(have_effect($effect[Consumed By Fear]) > 0)
+		{
+			uneffect($effect[Consumed By Fear]);
+			council();
+		}
+		return false;
+	}
+	if (in_koe())
+	{
+		if(item_amount($item[Digital Key]) == 0 && internalQuestStatus("questL13Final") == 5)
```

Comment:
I think moving logic that only runs at the tower door outside of the codeblock that runs right before unlocking the tower door is more confusing than leaving KoE-specific code at the tower door unlock section. I suggest moving this KoE-specific logic back to where it was.

Reply:
Before, when looking at that function, i just though "oh, koe must not use tower keys". it came as a surprise to find out that there was code for getting the koe key right before the door.
Additionally this is a function called "get key" and it is run right before the tower door.

it is actually kinda redundant for the tower door to go:
getkey()
code to get key in specific paths

instead of just going 
getkey()
and for that function to handle specific paths stuff.